### PR TITLE
Add a basic testing infrastructure and the bulk of the backfilled tests.

### LINKED ISSUE #19: Implement some proper unit tests, using @skylib or some other infrastructure

### PR DESCRIPTION
**Pull request by cgruber**
Add a basic testing infrastructure and the bulk of the backfilled tests.  This is patterned on skylib but simpler, and with less hooks into the infrastructure.  Also add tests for sets, utils, and poms bzl files. This also includes a few refactorings and renames as well as a few edge-cases fixed which were caught by the new tests.  Hooray for testing.

This is structured as it is, largely to avoid any users of the tooling to accidentally pull in the test code (since it runs in the analysis phase).

This fixes #19 and #21

### CODE REVIEW

**Comment by rjrjr** [comment]
I saw the tweet about this before anything else. But I'm definitely not having any trouble focusing.

File: maven/sets.bzl [hunk 1]
```diff
@@ -1,51 +1,71 @@
...
+implementation.  Generally only use dictionaries created by sets.new() because values are normalized. Using
+dictionaries from other sources may result in equality failing, and other odd behavior.
 """
 
 _UNDEFINED = "__UNDEFINED__"
-_EMPTY = "__EMPTY__"
+_EMPTY = "__EMPTY__"  # Check when changing this to keep in sync with sets.bzl
+_SET_DICTIONARY_KEY = "_____SET_DICTIONARY_KEY______"
 
 def _contains(set_dict, item):
```
Comment (rjrjr) [inline maven/sets.bzl]: s/`set_dict`/`set`/?

File: maven/tests/paths_test.bzl [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+def filename_test(env):
+    asserts.equals(env, "blah", paths.filename("//foo/bar/blah"))
+    asserts.equals(env, "blah", paths.filename("bar/blah"))
+    asserts.equals(env, "blah", paths.filename("/bar/blah"))
+    asserts.equals(env, "blah", paths.filename("../bar/blah"))
+    asserts.equals(env, "blah", paths.filename("@maven//bar/blah"))
+
+# Roll-up function.
+def paths_test_suite():
+    test_suite("paths", tests = [filename_test]
```
Comment (rjrjr) [inline maven/tests/paths_test.bzl]: Missing `)`

Comment (rjrjr) [reply maven/tests/paths_test.bzl]: Looks like next step is a travis build :)

File: maven/tests/strings_test.bzl [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+def trim_test_custom(env):
+    asserts.equals(env, "foo", strings.trim(" aaa fooa \n", "\n a"))
+
+def contains_test(env):
+    asserts.true(env, strings.contains("foobarbash", "bar"))
+    asserts.false(env, strings.contains("foobarbash", "Bar"))
+
+# Roll-up function.
+def strings_test_suite():
+    test_suite("strings", tests = [trim_test_custom, trim_test_default, contains_test])
```
Comment (rjrjr) [inline maven/tests/strings_test.bzl]: Will you know when you forget to add a test to these lists?

File: maven/tests/paths_test.bzl (hunk 2 above)
Comment (cgruber) [reply maven/tests/paths_test.bzl]: Wait... how did that even parse? o.O

Comment (cgruber) [reply maven/tests/paths_test.bzl]: Yeesh.  Travis build is right.  Whoops. I thought I had run it before checking that last bit in. :/

File: maven/tests/strings_test.bzl (hunk 3 above)
Comment (cgruber) [reply maven/tests/strings_test.bzl]: Nope.  I'm trying to think of a decent way to test for that.  Haven't yet. Possibly just parsing the files, and enforcing that any test ends in _test (and renaming a few of these), and that it's in these lists.  But there's some turtles all the way down situations here, and there's no reflective power to list all the functions in the file that I know of.

File: maven/sets.bzl (hunk 1 above)
Comment (cgruber) [reply maven/sets.bzl]: Done

File: maven/tests/paths_test.bzl (hunk 2 above)
Comment (cgruber) [reply maven/tests/paths_test.bzl]: Ok - Travis added to this PR, because why not.
